### ISSUE TITLE
Throw an exception inside coroutine launched with Dispatchers.Default doesn't crash the app

### ISSUE THREAD

**Issue by romanandreyvich** [Kotlin/kotlinx.coroutines#2398]
I throw an exception inside coroutine launched with Dispatchers.Default and it doesn't crash the app (just print stack trace in console). Works fine if I use Dispatchers.Main instead of Dispatchers.Default.

Here is an example project: https://drive.google.com/file/d/1NzwMU7QStgQuKBNgty1HZei1ZZvQG5Ys/view?usp=sharing

Kotlin: 1.4.20-RC
Coroutines: 1.3.9-native-mt
iOS: 14.1 (iPhone 11, simulator)

**Comment by Thomas-Vos** [comment]
I am running into this issue as well. Exceptions on `Dispatchers.Default` only print the stack trace but do not crash the app. It seems that an unhandled exception hook that was set using [`setUnhandledExceptionHook`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native/set-unhandled-exception-hook.html) also does not get called. I use that to report exceptions to Firebase, but as it does not get called, that does not work. Because of that it is very hard to fix crashes thrown from `Dispatchers.Default`. I am currently trying to work around this by using `CoroutineExceptionHandler`, that appears to be able to catch the exception. Would really appreciate a solution for this issue.

**Comment by Daniel-Jacob** [comment]
Is this only an issue on native? Has there been any issues on jvm with this?

**Comment by slipdef** [comment]
@qwwdfsad any update on this issue?

**Comment by qwwdfsad** [comment]
We'll try to fix it in the upcoming release

**Comment by slipdef** [comment]
@qwwdfsad any update on this? It looks like a serious issue - exception in coroutine doesn't crash an app as expected.

**Comment by Thomas-Vos** [comment]
Note that `setUnhandledExceptionHook` does not catch these exceptions. That could be a problem if you would like to log these exceptions to for example Firebase Crashlytics.

I am using the following code as a workaround so exceptions on background threads are also logged to Crashlytics:

```kt
@SharedImmutable
private val defaultExceptionHandler = CoroutineExceptionHandler { context, throwable ->
    throwable // TODO: do something with the throwable, for example log it to Firebase Crashlytics
    throw throwable
}

// Somewhere else in your code
fun main() {
    val dispatcher = Dispatchers.Default + defaultExceptionHandler
    GlobalScope.launch(dispatcher) {
        throw IllegalStateException()
    }
}
```

**Comment by qwwdfsad** [comment]
Fixed in `1.5.1-native-mt`. For an additional explanation of the behaviour see https://github.com/Kotlin/kotlinx.coroutines/pull/2804#discussion_r667930024

**Comment by Thomas-Vos** [comment]
Thanks, can confirm all is working correctly now.
